[Add] BaseUITests

DalioKim opened the pull request:
# 주요 변경사항
	- SnapshotTesting 라이브러리를 이용한 UIXCTestCase 컨테이너 클래스 BaseUITests 구현
	- BaseUITests 상속하여 AuthenticationUITests 구현
	- 설명: 기준이 되는 UI 레퍼런스 이미지를 생성하고, 그 레퍼런스와 새로 수정한 코드의 이미지를 비교할 수 있습니다.

# 중점적으로 봐줬으면 하는 부분
	- 분리된 계산속성과 함수의 적합성 
	- BaseUITests 사용성 및 간편성
	- 네이밍

# 추가 작업 예정사항
	- BDD 적용 예정

DalioKim committed:
FEAT: InputsValidator TDD 추가

1.InputsValidatorProtocol 추가 및 AuthenticationViewModel에 적용
2. MocksInputsValidatorProtocol, AuthenticationPresenterTests 추가
3. Validated_ShouldFail 테스트 케이스 구현 중

DalioKim committed:
FEAT:
1) 패스워드 유효성 테스트 추가
2) 성공 및 실패 샘플 생성 로직 구현
3) 네이밍 변경
4) 네임스페이스에 리터럴 상수 선언

DalioKim committed:
REFACTOR: 네이밍 변경 및 불필요한 코드 제거

DalioKim committed:
REFACTOR: 중복코드 제거, 네이밍 변경, 코드 구조 변경

1. BaseInputsValidator 클래스 선언 및 상속 통해서 vm과 mock의 중복 코드 제거
2. mock 클래스 네이밍 변경
3. AuthenticationPresenterTests private 부분 분리

DalioKim committed:
FEAT: AuthenticationView UI TDD

1. ViewInspector 라이브러리 추가 및 링크 설정
2. TDD를 위한 AuthenticationView 초기화 메서드 추가, id 설정, 프로퍼티 초기값 변경
3. submitButton 사용하지 않는 UI 로직 제거
4. AuthenticationUITests 로그인 여부를 알려주는 state 프로퍼티에 따른 UI 테스트 구현
5. 기타 -  AuthenticationPresenterTests 네임스페이스 추가
6. 기타 - MeasurementViewUITests 파일 경로변경

DalioKim committed:
FEAT: SnapshotTesting UI 테스트 추가
1. SnapshotTesting 라이브러리 추가 및 ViewInspector 임시 삭제
2. Test 번들경로 병합
3. DeviceModel, BaseUITests 등 UI테스트 추상화
3. AuthenticationUITests에 SnapshotTesting 적용

DalioKim committed:
REFACTOR: BaseUITests
1. 프로퍼티 및 사용법 주석으로 추가
2. setUpWithError 로직 수정
3. isRecord 스냅샷 생성인지 비교인지 구분하는 프로퍼티 추가
4. 파일명이 달라지는 이유로  verifySnapshot과 assertSnapshot 모두 하나의 메서드에서 isRecord로 구분해서 호출하는 방식으로 수정
5. 프로퍼티 및 메서드 네이밍 수정
6. AuthenticationUITests에 적용

hanjaesueng commented on GatheRunnerTests/Utils/BaseUITests.swift:
```diff
@@ -0,0 +1,157 @@
+//
+//  BaseUITests.swift
+//  GatheRunnerTests
+//
+//  Created by 김동현 on 2022/10/07.
+//
+
+import SnapshotTesting
+import SwiftUI
+import XCTest
+@testable import GatheRunner
+
+// MARK: - BaseUITests
+
+class BaseUITests: XCTestCase {
+
+    // MARK: Internal
+
+    /// Inherited classes must be overridden Computed Properties.
+    ///
```
질문
-----------------------------------------------
BaseUITests의 의도와 용도를 알고싶습니다

hanjaesueng replied:
그리고 SnapshotTesting 은 휴대폰 스크린샷 테스트 용도인가요???

DalioKim replied:
> ## 질문
> BaseUITests의 의도와 용도를 알고싶습니다

UI테스팅에서 기본적으로 사용되는 테스트들을 정의해서 재사용할 의도로 만들었습니다.
아래에서 설명드리듯이 기준 이미지와 지금 작성한 코드의 이미지를 비교하는 테스팅인데 
기종별 아이폰, 설정별(다크모드,가로모드)등의 테스트는 중복적으로 실행되기 때문에 이런 부분들을 매 UI 테스팅때마다 구현하지 않고, BaseUITests를 이용해 테스팅할 뷰와 기종, 설정만 지정해주면 테스팅 해 줄 수 있게 만들어봤습니다.

DalioKim replied:
> 그리고 SnapshotTesting 은 휴대폰 스크린샷 테스트 용도인가요???

네넵 기준 PNG 파일이 있으면, 현재 구현된 UI의 스크린샷과 비교해줍니다. 다만 픽셀단위로 체크하기때문에 한 부분만 틀려도 일치하지 않는다고 뜨기때문에, 디자이너가 바로 넘겨준 제안이미지를 기준으로 사용하긴 힘들고, 디자이너가 제안한걸 토대로 최대한 유사하게 구현한 UI를 기준 이미지로로 사용하고, 추후에 여러 변경이 있을때 기준 이미지를 기준으로 바뀐 코드와 비교할때 용이합니다.

hanjaesueng replied:
그러면 디자인이 있을때 1 픽셀까지 정확히 일치하게 구현해야 통과 가능한가요??

DalioKim replied:
네 맞습니다. 
https://blog.banksalad.com/tech/test-in-banksalad-ios-2/ 뱅크샐러드에서 사용하고 방식인데 한번 참고해보시면 좋을 것 같습니다.

wh5929 commented on GatheRunner/Presentaion/Authentication/Protocols/InputsValidatorProtocol.swift:
```diff
@@ -0,0 +1,62 @@
+public class BaseInputsValidator: ObservableObject, InputsValidatorProtocol {
+
+    // MARK: Lifecycle
+
+    init() {
+        bindValidation()
+    }
+
+    // MARK: Public
+
+    @Published public var isAuthValid = false
+    @Published public var isEmailValid = false
+    @Published public var isPasswordValid = false
+    @Published public var isInputsValid = false
+
+    public var cancelBag = Set<AnyCancellable>()
+
+    // MARK: Internal
+
+    @Published var email = ""
+    @Published var password = ""
+
+    // MARK: Private
+
+    private func bindValidation() {
+        $email
+            .compactMap { $0 }
+            .sink { [weak self] in
+                self?.isEmailValid = $0.isEmailValid
+            }
+            .store(in: &cancelBag)
+
+        $password
+            .compactMap { $0 }
+            .sink { [weak self] in
+                self?.isPasswordValid = $0.isPasswordValid
+            }
+            .store(in: &cancelBag)
+    }
+}
```
### 의견

프로토콜과 그 구현부는 파일을 분리 하시는게 어떠신지요?
InputsValidatorProtocol.swift 파일에 BaseInputsValidator 클래스가 있으니 어색한것 같습니다.

DalioKim replied:
프로토콜과 class 혹은 struct를 한 파일에 두는게 어색하진 않은 것 같습니다.
[예시](https://github.com/nalexn/clean-architecture-swiftui/blob/master/CountriesSwiftUI/Repositories/CountriesWebRepository.swift)

예시링크를 보시면 프토토콜이 한 타입의 청사진으로써 같은 파일에 위치하는걸 확인하실 수 있습니다.

wh5929 replied:
> 

정의부와 구현부를 따로 분리하지 않나보네요.
참고하겠습니다.

DalioKim replied:
이건 저도 좀 더 확인해보겠습니다. 저렇게 배치한 이유까지는 아직 파악하거나 이해한건 아니어서요.

wh5929 commented on GatheRunnerTests/Authentication/Mocks/MocksInputsValidator.swift:
```diff
@@ -0,0 +1,18 @@
+        } else {
+            password = exampe
+        }
+    }
```
### 의견

exampe  오타 같습니다.

DalioKim committed:
FIX: 오타 수정 - MocksInputsValidator

REFACTOR:

AuthenticationView & AuthenticationViewModel
1. 불필요한 코드 삭제
2. validatedInputs 계산속성으로 변경

MocksInputsValidator
1. InputType 네임스페이스 정의
2. didValidation 메서드 네이밍, 파라미터, 로직 변경

AuthenticationPresenterTests
1. Option 네임스페이스 정의
2. example관련 메서드 병합 및 메서드 네이밍(명사형), 로직 변경

View+ToVc
1. convertedVC로 네이밍 변경

DalioKim commented:
# 주요 변경사항
	- 오타 수정 
	- 네임스페이스 정의 및 네임 스페이스를 활용한 로직 변경
	- 네이밍 변경

# 중점적으로 봐줬으면 하는 부분
	- 네임스페이스 위치
	- 변경된 로직
	- 네이밍
